File: game-theory/java/WGME.java
import java.util.*;
import java.io.*;
public class Main{
	public static String str;
	public static int grundy[] = new int[100000];
	public static int solve(int mask){
		if(grundy[mask]!=(-1))
			return grundy[mask];
		if(mask==0)
			return 0;
		int hash[] = new int[26];
		for(int i=0;i<str.length();i++)
			if((mask & (1<<i))!=0)
				hash[str.charAt(i)-'a']++;
		int flag = 0;
		for(int i=0;i<26;i++)
			if(hash[i]>1)
				flag = 1;
		if(flag==0)
			return 0;
		ArrayList<Integer> number = new ArrayList<Integer>();
		for(int i=0;i<str.length();i++)
			if((mask & (1<<i))!=0){
				mask = mask^(1<<i);
				int temp = solve(mask);	
				number.add(temp);
				mask = mask^(1<<i);	
			}
		Collections.sort(number);
		int small = 0;
		for(int i=0;i<number.size();i++){
			int temp = number.remove(0);
			if(temp>small)
				break;
			else
				small++;
		}
		grundy[mask] = small;
		return small;
	}
	public static void main(String[] args)throws IOException{
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		int t = Integer.parseInt(br.readLine());
		while(t-->0){
			str = br.readLine().trim();
			Arrays.fill(grundy,-1);
			int ret = solve((1<<str.length())-1);
			if(ret==0)
				System.out.println("Watan");
			else
				System.out.println("Karan");
		}
	}
}

File: game-theory/C++/WGME.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <string>

using namespace std;

#define pb push_back

vector <int> grundy;
string inp;

int solve(int mask)
{
	if (mask == 0)							// no possible moves
		return 0;

	if (grundy[mask] != -1)					// check if already calculated
		return grundy[mask];

	int hash[26] = {}, flag = 0;
	for (int i = 0; i < inp.length(); i++)	// count occurrence of each character
	{
		if (mask & (1 << i))
			hash[inp[i] - 'a']++;
	}

	for (int i = 0; i < 26; i++)
	{
		if (hash[i] > 1)					// check for duplicates
		{
			flag = 1;
			break;
		}
	}

	if (flag == 0)							// no duplicates means game over
	{
		grundy[mask] = 0;
		return 0;
	}

	vector <int> list;
	for (int i = 0; i < inp.length(); i++)
	{
		if (mask & (1 << i))				// check if element exists and go to possible game states
			list.pb(solve(mask - (1 << i)));
	}

	sort(list.begin(), list.end());
	int mex = 0;
	for (auto i : list)						// finding mex
	{
		if (mex < i)
			break;
		
		if (mex == i)
			mex++;
	}

	grundy[mask] = mex;
	return grundy[mask];
}

int main()
{
	int T;
	cin >> T;

	int mask;
	for (int i = 0; i < T; i++)
	{
		cin >> inp;

		mask = (1 << inp.length());
		mask--;

		vector <int> temp((1 << inp.length()), -1);
		//initialize grundy to -1
		grundy = temp;

		solve(mask) == 0 ? cout << "Watan\n" : cout << "Karan\n";
	}

	return 0;
}